### PR TITLE
Update swiss_species Encoding Data Source (JSON) #999

### LINKED ISSUE #999: Update swiss_species Encoding Data Source (JSON)
## Description
The goal is to replace the Swiss Species Registry (SSR) csv with a SSR Checklist on ChecklistBank. Currently this SSR-Checklist is **not** ready/available.

In this issue the existing swiss_species_registry.csv will be replaced with an updated 'Swiss Species Registry' as JSON that mimics the new Swiss Species Registry.

(The JSON file is stored on our [Google Drive](https://drive.google.com/drive/u/0/folders/1Fryd5D18nivyzNukxr3EoJ-X_K5ikf_d))

## Pre-Requisites
- [x] The new 'Swiss Species Registry' JSON offers als needed attributes with correct naming
- [x] Exact attribute mapping (DAGI <> SSR)

## Desired
- Create a new and parallel table containing the new Swiss Species Registry (SSR)
- Populate new SSR table with Swiss Taxon Information
- For Encoding use scientificName instead of TaxonID
- Encoding Step: Populate `taxonIdCH`, `acceptedNameUsage`,`swissSpeciesCenter`, `swissSpeciesRegisteredAt` and `swissSpeciesRegistered`
- Catch if >1 result is received from SSR
- Translation / Helper for Data Center Mapping

## Accepted if
- [x] Swiss Species Encoding is based on updated SSR table
- [x] Base for Swiss Species Encoding is `scientificName` and not `taxonID` anymore
- [x] Attribute `taxonIdCH` is being populated with `result.id`
- [x] Attribute `acceptedNameUsage` is being populated with `result.usage.label` IF `result.usage.status` = "accepted"
- [x] Attribute `acceptedNameUsage` is being populated with `result.usage.accepted.name.label` IF `result.usage.status` = "synonym"
- [x] Attribute `swissSpeciesCenter` is being populated
- [x] Attribute `swissSpeciesRegisteredAt` is being populated
- [x] Attribute `swissSpeciesRegistered` is being populated
- [x] When encoding of any attribute fails set Encoding Status to 'Incomplete'
- [x] If encoding leads to >1 result set Encoding Status to Incomplete (Add Error: Found duplicate in Swiss Species Registry.)

### PR DESCRIPTION
**Pull request by sidler92**
# Pull Request

## Description

- safe json file as ndjson using this command: 
`jq -c 'to_entries[]' swiss_species_registry.json > swiss_species_registry.ndjson`
-  Add new `swiss_species_registry` table and populate it using new importer using the newly provided ndjson file
- Swiss Species Registry encoding step now uses the `scientificName` instead of `taxonID`
- Output values of the SSR encoding step are as defined by the issue #999

## Related Issue

fixes #999

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Performance improvement
- [x] Test update
- [ ] Build/CI update
- [ ] Other (please describe):

## License Compliance

- [x] My changes comply with the GNU AGPL-3.0 license requirements
- [x] If my changes involve network interaction, I've ensured users can access the corresponding source code

### CODE REVIEW

File: priv/initialize/catalogs/swiss_species_registry.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+version https://git-lfs.github.com/spec/v1
```
Comment (cloedu87) [inline priv/initialize/catalogs/swiss_species_registry.json]: you can also use:
```bash
jq -c . < swiss_species_registry.json > mini.json
```
which significantly reduces the file size... like:

```bash
❯ jq -c . < swiss_species_registry.json > mini.json
❯ ll
30M Jan  7 12:49 mini.json
85M Jan  7 12:35 swiss_species_registry.json
```

or convert it to a line based, minfied json, like suggested above:

```bash
jq -c 'to_entries[]' swiss_species_registry.json > swiss_species_registry.ndjson
```

File: test/data_aggregator/taxonomy/catalogs/swiss_species_registry_importer_test.exs [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+      assert entry.scientific_name == "Parasyrisca vinosa (Simon, 1878)"
+      assert entry.taxon_id_ch == "10000"
+      assert entry.accepted_name_usage == "Parasyrisca vinosa (Simon, 1878)"
+      assert entry.center == :infofauna
+      assert entry.rank == "species"
+      assert entry.status == "accepted"
+    end
+
+    test "imports multiple accepted species entries" do
```
Comment (cloedu87) [inline test/data_aggregator/taxonomy/catalogs/swiss_species_registry_importer_test.exs]: this is a bit testing the same as the first test.

if you agree on testing with the real data, I would rather add another test, which actually verifies the amount of imported lines to be sure, the data is imported correct and complete when we re-deploy. Useful for dependency updates and changes on the modules for importing.

```elixir
# dummy-ish...
assert Ash.count!(SwissSpeciesRegistry) == 424242
```
you could also test if the data fields in the database are properly filled and not empty by grouping results etc...

File: test/data_aggregator/taxonomy/catalogs/swiss_species_registry_importer_test.exs [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+  alias DataAggregator.Taxonomy.Catalogs.SwissSpeciesRegistry
+  alias DataAggregator.Taxonomy.Catalogs.SwissSpeciesRegistryImporter
+
+  doctest SwissSpeciesRegistryImporter, import: true
+
+  @fixtures_path "test/support/fixtures/files"
+
+  describe "import_from_json/1" do
+    test "imports accepted species entries correctly" do
+      path = Path.join(@fixtures_path, "swiss_species_registry_test.json")
```
Comment (cloedu87) [inline test/data_aggregator/taxonomy/catalogs/swiss_species_registry_importer_test.exs]: I know the file is huge, but we should consider testing with the "real" data for some cases.
first to verify if (all of) the file is actually readable in case it changes and second to be aware of library changes during dependency updates which might affect edge cases in the data itself.

if you agree on testing/checking the data itself is of value instead of only testing the import logic along with the mocked data.

if this makes sense for you, I would move the import logic to a `setup_all` clause, so it gets executed only once and not for each test:

```elixir
describe "import_from_json/1" do
  setup_all do
    assert :ok = SwissSpeciesRegistryImporter.import_from_json(path)

    {:ok, []}
  end
end
```
instead of doing this or if the test is running very long, you could also consider setting a `@tag swiss_species_registry: true` to the tests who are actually test the real-data and execute them only on the CI and in case the file changed. like:

```yaml
# cicd.yml
on:
  push:
    paths:
      - "path/to/registry-file.json"
```

File: lib/data_aggregator/taxonomy/catalogs/swiss_species_registry_importer.ex [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+  }
+
+  @doc """
+  Import Swiss Species Registry from a JSON file.
+  """
+  @spec import_from_json(String.t()) :: :ok
+  def import_from_json(path) do
+    Logger.info("[swiss_species_registry_importer] Starting import from #{path}")
+
+    path
```
Comment (cloedu87) [inline lib/data_aggregator/taxonomy/catalogs/swiss_species_registry_importer.ex]: as this gets executed on a pod during deployment and they have rather low memory it might be an option to stream the json data.

one option here could be `jaxon`

```elixir
# ~
path
|> File.stream!()
|> Jaxon.Stream.from_enumerable()
|> Jaxon.Stream.query([:root, :all])
|> Stream.each(&import_entry/1)
|> Stream.run()
```

or convert it to a `ndjson`

```bash
jq -c 'to_entries[]' swiss_species_registry.json > swiss_species_registry.ndjson
```
to get a line-based json structure like this:

```ndjson
{"key":"Parasyrisca vinosa","value":{"result":[{"id":"infofauna:10000","classification":[{"rank":"genus","label":"Parasyrisca Schenkel, 1963"},{"rank":"family","label":"Gnaphosidae Banks, 1892"},{"rank":"order","label":"Araneae"},{"rank":"class","label":"Arachnida"},{"rank":"subphylum","label":"Chelicerata"},{"rank":"phylum","label":"Arthropoda"},{"rank":"kingdom","label":"Animalia"},{"rank":"domain","label":"Eukaryota (Chatton, 1925) Whittaker & Margulis, 1978"}],"usage":{"name":{"rank":"species","issues":null},"status":"accepted","label":"Parasyrisca vinosa (Simon, 1878)"}}]}}
{"key":"Echemus angustifrons","value":{"result":[{"id":"infofauna:10001","classification":[{"rank":"genus","label":"Echemus Simon, 1878"},{"rank":"family","label":"Gnaphosidae Banks, 1892"},{"rank":"order","label":"Araneae"},{"rank":"class","label":"Arachnida"},{"rank":"subphylum","label":"Chelicerata"},{"rank":"phylum","label":"Arthropoda"},{"rank":"kingdom","label":"Animalia"},{"rank":"domain","label":"Eukaryota (Chatton, 1925) Whittaker & Margulis, 1978"}],"usage":{"name":{"rank":"species","issues":null
```

then you could stream it with elixir onboard equipment:
```elixir
path 
  |> Path.expand()
  |> File.stream!()
  |> Stream.map(&Jason.decode!/1)
  |> Enum.each(&import_entry/1)

# resolve then line by line like this:
def import_entry(%{"key" => scientific_name, "value" => data}) do
  # do magic here 🪄 
end
```

File: test/data_aggregator/taxonomy/catalogs/swiss_species_registry_importer_test.exs [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    test "logs warning for entries with empty results" do
+      # Create a temporary JSON file with empty results
+      tmp_path = Path.join(System.tmp_dir!(), "empty_results_test.json")
+
+      json_content =
+        Jason.encode!(%{
+          "Unknown species" => %{"result" => []}
+        })
+
+      File.write!(tmp_path, json_content)
```
Comment (cloedu87) [inline test/data_aggregator/taxonomy/catalogs/swiss_species_registry_importer_test.exs]: instead of writing files here you could make a small refactoring and make the `SwissSpeciesRegistryImporter.import_from_json/1` accept a path OR a json-structure (or directly a map or list of maps or stream of maps)

**Review by cloedu87 — CHANGES_REQUESTED** [review]
looks good, some stuff to consider in the comments...

File: priv/initialize/catalogs/swiss_species_registry.json (hunk 1 above)
Comment (sidler92) [reply priv/initialize/catalogs/swiss_species_registry.json]: I opted to use the ndjson approach.
This leads to the file being smaller again -> removed the git-lfs stuff again (this will safe us some headaches in the future)
Thanks for the tip

File: lib/data_aggregator/taxonomy/catalogs/swiss_species_registry_importer.ex (hunk 4 above)
Comment (sidler92) [reply lib/data_aggregator/taxonomy/catalogs/swiss_species_registry_importer.ex]: I stream the data now in chunks

File: test/data_aggregator/taxonomy/catalogs/swiss_species_registry_importer_test.exs (hunk 3 above)
Comment (sidler92) [reply test/data_aggregator/taxonomy/catalogs/swiss_species_registry_importer_test.exs]: I refactored the importer, to use bulk_create, I could reduce the runtime with it from 2 minutes to 8 seconds =)

-> I added a test that tests the real data

File: test/data_aggregator/taxonomy/catalogs/swiss_species_registry_importer_test.exs (hunk 5 above)
Comment (sidler92) [reply test/data_aggregator/taxonomy/catalogs/swiss_species_registry_importer_test.exs]: I removed these tests,
Now I just have 2 ndjson test files:
one for valid data tests, and if the fields are filled correctly
one with partly invalid data, to check correct error handling

File: test/data_aggregator/taxonomy/catalogs/swiss_species_registry_importer_test.exs (hunk 2 above)
Comment (sidler92) [reply test/data_aggregator/taxonomy/catalogs/swiss_species_registry_importer_test.exs]: refactored tests

**Review by cloedu87 — APPROVED** [review]
fantastico🤌
🥳
